lib/ephy-sync-utils.c

```diff
@@ -0,0 +1,185 @@
+/* -*- Mode: C; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/*
+ *  Copyright © 2017 Gabriel Ivascu <ivascu.gabriel59@gmail.com>
+ *
+ *  This file is part of Epiphany.
+ *
+ *  Epiphany is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Epiphany is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Epiphany.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "config.h"
+#include "ephy-sync-utils.h"
+
+#include <stdio.h>
+#include <string.h>
+
+#define SYNC_ID_LEN 12
+
+static const char hex_digits[] = "0123456789abcdef";
+
+char *
+ephy_sync_utils_encode_hex (const guint8 *data,
+                            gsize         data_len)
+{
+  char *encoded;
+
+  g_return_val_if_fail (data, NULL);
+
+  encoded = g_malloc (data_len * 2 + 1);
+  for (gsize i = 0; i < data_len; i++) {
+    guint8 byte = data[i];
+
+    encoded[2 * i] = hex_digits[byte >> 4];
+    encoded[2 * i + 1] = hex_digits[byte & 0xf];
+  }
+  encoded[data_len * 2] = 0;
+
+  return encoded;
+}
+
+guint8 *
+ephy_sync_utils_decode_hex (const char *hex)
+{
+  guint8 *decoded;
+
+  g_return_val_if_fail (hex, NULL);
+
+  decoded = g_malloc (strlen (hex) / 2);
+  for (gsize i = 0, j = 0; i < strlen (hex); i += 2, j++)
+    sscanf (hex + i, "%2hhx", decoded + j);
+
+  return decoded;
+}
+
+static void
+base64_to_base64_urlsafe (char *text)
+{
+  g_assert (text);
+
+  /* Replace '+' with '-' and '/' with '_' */
```

Comment:
@mcatanzaro  / and + are inappropriate for URLs and file systems paths so to make a base64 string safe there's a convention to replace / with _ and + with -. There's some info about that here: https://en.wikipedia.org/wiki/Base64#URL_applications. Other libraries have APIs that encode directly to a safe base64 string (such as Python's `base64.urlsafe_b64encode`), but GLib doesn't so I had to do it myself.